jxl/src/render/stages/noise.rs

```diff
@@ -0,0 +1,357 @@
+            let red_noise = noise_strength_r
+                * (k_rgn_corr * addit_rnd_noise_red + k_rg_corr * addit_rnd_noise_correlated);
+            let green_noise = noise_strength_g
+                * (k_rgn_corr * addit_rnd_noise_green + k_rg_corr * addit_rnd_noise_correlated);
+            let rg_noise = red_noise + green_noise;
+            row[0][x] += ytox * rg_noise + red_noise - green_noise;
+            row[1][x] += rg_noise;
+            row[2][x] += ytob * rg_noise;
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        error::Result,
+        features::noise::Noise,
+        image::Image,
+        render::{
+            stages::noise::{AddNoiseStage, ConvolveNoiseStage},
+            test::make_and_run_simple_pipeline,
+            RenderPipelineStage,
+        },
+        util::test::assert_almost_eq,
+    };
+    use test_log::test;
+
+    #[test]
+    fn convolve_noise_uses_channels() -> Result<()> {
+        let stage = ConvolveNoiseStage::new(1);
+        assert!(
+            !stage.uses_channel(0),
+            "shouldn't use channel < first_channel"
+        );
+        for c in 1..=3 {
+            assert!(
+                stage.uses_channel(c),
+                "should use first_channel <= channel < first_channel + 3"
+            );
+        }
+        assert!(
+            !stage.uses_channel(4),
+            "shouldn't use channel > first_channel + 3"
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn convolve_noise_process_row_chunk() -> Result<()> {
+        let input_c0: Image<f32> = Image::new_range((2, 2), 0.0f32, 1.0f32)?;
+        let input_c1: Image<f32> = Image::new_range((2, 2), 0.0f32, 1.0f32)?;
+        let input_c2: Image<f32> = Image::new_range((2, 2), 0.0f32, 1.0f32)?;
+        let stage = ConvolveNoiseStage::new(0);
+        let output: Vec<Image<f32>> =
+            make_and_run_simple_pipeline(stage, &[input_c0, input_c1, input_c2], (2, 2), 256)?.1;
+        let rect = output[0].as_rect();
+        assert_almost_eq!(rect.row(0)[0], 7.2f32, 1e-6f32);
+        assert_almost_eq!(rect.row(0)[1], 2.4f32, 1e-6f32);
+        assert_almost_eq!(rect.row(1)[0], -2.4f32, 1e-6f32);
+        assert_almost_eq!(rect.row(1)[1], -7.2f32, 1e-6f32);
+        Ok(())
+    }
+
+    #[test]
+    fn add_noise_uses_channels() -> Result<()> {
+        let stage = AddNoiseStage::new(
+            Noise {
+                lut: [
+                    0.0f32, 0.0f32, 0.0f32, 0.0f32, 0.0f32, 0.0f32, 0.0f32, 0.0f32,
+                ],
+            },
+            1,
+        );
+        assert!(
+            !stage.uses_channel(0),
+            "shouldn't use channel < first_channel"
+        );
+        assert!(stage.uses_channel(1), "should use channel == first_channel");
+        assert!(
+            !stage.uses_channel(4),
+            "shouldn't use channel > first_channel + 3"
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn add_noise_process_row_chunk() -> Result<()> {
```

Comment:
I'd prefer to test specific situations for which we can easily know what is supposed to happen.

For example, we could have a constant noise LUT, 0 values in the first two noise channels, and check that adding noise does nothing on color channel 0, and adds the 0.0034375*noise_channel_2 to the second and third color channels.

Another reasonable test is to verify that the applied noise does not depend on the third color channel.

Some other tests with the noise LUT below should also be relatively easy, as in that setup noise.eval(x) ought to just be a linear function in the appropriate range.

Reply:
Added TODO(mo271) for this as per Discord discussion.